server/build/webpack/utils.js

```diff
@@ -6,36 +6,38 @@ const glob = promisify(globModule)
 
 const nextPagesDir = path.join(__dirname, '..', '..', '..', 'pages')
 
-export async function getPages (dir, {dev, isServer, pageExtensions}) {
-  const pageFiles = await getPagePaths(dir, {dev, isServer, pageExtensions})
-
-  return getPageEntries(pageFiles, {isServer, pageExtensions})
+export async function getPages (dir, {dev, isServer, pageExtensions, rootPaths = ['.']}) {
+  const pagePathsByRoot = await getPagePaths(dir, {dev, isServer, pageExtensions, rootPaths})
+  return getMultiRootPageEntries(pagePathsByRoot, {isServer, pageExtensions})
 }
 
-export async function getPagePaths (dir, {dev, isServer, pageExtensions}) {
-  let pages
-
-  if (dev) {
-    // In development we only compile _document.js, _error.js and _app.js when starting, since they're always needed. All other pages are compiled with on demand entries
-    // _document also has to be in the client compiler in development because we want to detect HMR changes and reload the client
-    pages = await glob(`pages/+(_document|_app|_error).+(${pageExtensions})`, { cwd: dir })
-  } else {
-    // In production get all pages from the pages directory
-    pages = await glob(isServer ? `pages/**/*.+(${pageExtensions})` : `pages/**/!(_document)*.+(${pageExtensions})`, { cwd: dir })
-  }
-
-  return pages
+export async function getPagePaths (dir, {dev, isServer, pageExtensions, rootPaths = ['.']}) {
+  // In development we only compile _document.js, _error.js and _app.js when starting, since they're always needed. All other pages are compiled with on demand entries
+  // _document also has to be in the client compiler in development because we want to detect HMR changes and reload the client.
+  // In production get all pages from the pages directory.
+  const filePattern = dev
+    ? `pages/+(_document|_app|_error).+(${pageExtensions})`
+    : isServer
+      ? `pages/**/*.+(${pageExtensions})`
+      : `pages/**/!(_document)*.+(${pageExtensions})`
+
+  return Promise.all(rootPaths.map(async root => {
+    const pages = await glob(`${root}/${filePattern}`, { cwd: dir })
+    return { root, pages }
+  }))
 }
 
 // Convert page path into single entry
-export function createEntry (filePath, {name, pageExtensions} = {}) {
+export function createEntry (filePath, {name, pageExtensions, root = '.'} = {}) {
   const parsedPath = path.parse(filePath)
-  let entryName = name || filePath
+  const relativePath = path.relative(root, filePath)
+  const parsedRelativePath = path.parse(relativePath)
+  let entryName = name || relativePath
 
   // This makes sure we compile `pages/blog/index.js` to `pages/blog.js`.
   // Excludes `pages/index.js` from this rule since we do want `/` to route to `pages/index.js`
-  if (parsedPath.dir !== 'pages' && parsedPath.name === 'index') {
-    entryName = `${parsedPath.dir}.js`
+  if (parsedRelativePath.dir !== 'pages' && parsedRelativePath.name === 'index') {
```

Comment:
Curious how would it play with a custom pages path.

Reply:
Does Next.js support custom page paths? It looked to be pretty hard-coded.